File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.py
'''
- populate minHeap first node from each list
- we only need the first node since we have references for the remaining nodes through that
- while that minHeap is valid, build the list
- repeat with the remaining indices
'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        dummy = ListNode()
        curr = dummy
        minHeap = []
        heapq.heapify(minHeap)

        for i in range(len(lists)):
            if lists[i]:
                head = lists[i]
                heapq.heappush(minHeap, [head.val, i, head])

        while minHeap:
            value, i, leastNode = heapq.heappop(minHeap)
            curr.next = leastNode
            curr = curr.next
            nextNode = leastNode.next
            
            if nextNode: 
                heapq.heappush(minHeap, [nextNode.val, i, nextNode])
        
        return dummy.next

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        map<Node*, Node*> nodeMap;
        Node* curr = head;
        nodeMap[NULL] = NULL;

        while (curr) {
            // Populate Map
            nodeMap[curr] = new Node(curr->val);

            // Move Pointer
            curr = curr->next;
        }

        Node* secondCurr = head;

        while (secondCurr) {
            Node* copiedNode = nodeMap[secondCurr];

            copiedNode->next = nodeMap[secondCurr->next];
            copiedNode->random = nodeMap[secondCurr->random];
        
            secondCurr = secondCurr->next;
        }

        return nodeMap[head];

    }
};
